File: com.se.simplicity/src/com/se/simplicity/vector/SimpleTranslationVectorf4.java
package com.se.simplicity.vector;

/**
 * <p>
 * A 4 dimensional translation vector that stores its data in a <code>float</code> array.
 * </p>
 * 
 * <p>
 * This <code>SimpleVectorf4</code> uses a 3D homogeneous coordinate system. It is of the following format:
 * </p>
 * 
 * <pre>
 * ----------------
 *  x | y | z | w |
 * ----------------
 * </pre>
 * 
 * @author simple
 */
public class SimpleTranslationVectorf4 extends SimpleVectorf4 implements TranslationVectorf
{
	/**
	 * The version of this class.
	 */
	private static final long serialVersionUID = 1L;
	
	/**
	 * <p>
	 * Creates an instance of <code>SimpleTranslationVectorf4</code>.
	 * </p>
	 */
	public SimpleTranslationVectorf4()
	{
		super();
	}
	
	/**
	 * <p>
	 * Creates an instance of <code>SimpleTranslationVectorf4</code>.
	 * </p>
	 * 
	 * @param f0 The x element of this <code>SimpleTranslationVectorf4</code>.
	 * @param f1 The y element of this <code>SimpleTranslationVectorf4</code>.
	 * @param f2 The z element of this <code>SimpleTranslationVectorf4</code>.
	 * @param f3 The w element of this <code>SimpleTranslationVectorf4</code>.
	 */
	public SimpleTranslationVectorf4(final float f0, final float f1, final float f2, final float f3)
	{
		super(f0, f1, f2, f3);
	}
	
	/**
	 * <p>
	 * Creates an instance of <code>SimpleTranslationVectorf4</code>.
	 * </p>
	 * 
	 * @param array An array containing the initial elements of this <code>SimpleTranslationVectorf4</code>.
	 */
	public SimpleTranslationVectorf4(final float[] array)
	{
		super(array);
	}
	
	@Override
	public float getW()
	{
		return (getArray()[3]);
	}

	@Override
	public float getX()
	{
		return (getArray()[0]);
	}

	public float getY()
	{
		return (getArray()[1]);
	}

	@Override
	public float getZ()
	{
		return (getArray()[2]);
	}

	@Override
	public void setX(final float x)
	{
		getArray()[0] = x;
	}

	@Override
	public void setY(final float y)
	{
		getArray()[1] = y;
	}

	@Override
	public void setZ(final float z)
	{
		getArray()[2] = z;
	}

	@Override
	public void translateX(final float x)
	{
		getArray()[0] += x;
	}

	@Override
	public void translateY(final float y)
	{
		getArray()[1] += y;
	}

	@Override
	public void translateZ(final float z)
	{
		getArray()[2] += z;
	}
}

File: simplicity/src/simplicity/vector/SimpleTranslationVectorf4.cpp
/*
 This file is part of The Simplicity Engine.

 The Simplicity Engine is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published
 by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

 The Simplicity Engine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with The Simplicity Engine. If not, see <http://www.gnu.org/licenses/>.
 */
#include "SimpleTranslationVectorf4.h"

namespace simplicity
{
    SimpleTranslationVectorf4::SimpleTranslationVectorf4() :
        SimpleVectorf4()
    {
    }

    SimpleTranslationVectorf4::SimpleTranslationVectorf4(const float f0, const float f1, const float f2, const float f3) :
        SimpleVectorf4(f0, f1, f2, f3)
    {
    }

    SimpleTranslationVectorf4::SimpleTranslationVectorf4(float* const array) :
        SimpleVectorf4(array)
    {
    }

    SimpleTranslationVectorf4::~SimpleTranslationVectorf4()
    {
    }

    float
    SimpleTranslationVectorf4::getW()
    {
        return (getArray()[3]);
    }

    float
    SimpleTranslationVectorf4::getX()
    {
        return (getArray()[0]);
    }

    float
    SimpleTranslationVectorf4::getY()
    {
        return (getArray()[1]);
    }

    float
    SimpleTranslationVectorf4::getZ()
    {
        return (getArray()[2]);
    }

    void
    SimpleTranslationVectorf4::setX(const float x)
    {
        getArray()[0] = x;
    }

    void
    SimpleTranslationVectorf4::setY(const float y)
    {
        getArray()[1] = y;
    }

    void
    SimpleTranslationVectorf4::setZ(const float z)
    {
        getArray()[2] = z;
    }

    void
    SimpleTranslationVectorf4::translateX(const float x)
    {
        getArray()[0] += x;
    }

    void
    SimpleTranslationVectorf4::translateY(const float y)
    {
        getArray()[1] += y;
    }

    void
    SimpleTranslationVectorf4::translateZ(const float z)
    {
        getArray()[2] += z;
    }
}
